### ISSUE TITLE
How to recieve images?

### ISSUE THREAD

**Issue by piyushmz** [eclipse-ecal/ecal-camera-samples#1]
I'm trying this program it's sending the images to a topic. Is there any code available for receiving the images?

The other way I'm trying is I've a camera driver of which is written in C & gstreamer for multiple cameras. I've added the publisher in C driver & I need to receive it in a cpp program.
Publisher in C
```
eCAL_Initialize(argc, argv, "e-multicam", eCAL_Init_Default);
pub = eCAL_Pub_New();
eCAL_Pub_Create(pub, "Hello", "base:std::string", "", 0);
...
for (cam = 0; cam < app_data->cameras_connected; cam++) {
if ((stream_data.dq_err & (1 << cam))) {
      /* 
      * Capture the frames in cap_ptr buffer.This buffer is passed around to display unit and 
      * image capture functions. Use this buffer for any algorithms (e.g: OpenCV)
      */
	cap_ptr[cam] = get_frame_virt(app_data, camera_buffer + cam, cam);
	eCAL_Pub_Send(pub, cap_ptr, *cap_ptr[cam], -1);
}
```
I just want to send `cap_ptr[cam]` buffer & receive it in my rec.cpp
cap_ptr is a struct as follows
```
struct cap_ptr {
	guint8 *cap_ptra;
	guint8 *cap_ptrb;
	guint8 *cap_ptrc;
	guint8 *cap_ptrd;
	guint8 *cap_ptre;
	guint8 *cap_ptrf;

};
```
Subscriber in CPP is same as `minimal_rec.cpp`

Thanks.

**Comment by KerstinKeller** [comment]
Hi @piyushmz,

eCAL is a framework to send data across processes. You can send any kind of data (e.g. a bytes array), or you can use serialization formats (e.g. Protobuf) on top of the binary communication.
This makes it easier to transfer the data, because between different processes, memory layout of data may be different (because of different memory layouts, due to compilers, flags or other reasons).

If you are using the C API of eCAL (see your sample), the layers on top of binary transport are not available. You can only send binary data.

However, your struct which you are trying to send, it contains **pointers** to data. These are addresses to memory, which only live in you sender application. They are more or less useless on receiver side, because the actual data does not exists on receiver side.

If you want to actually transfer the data, you need to define a struct that contains arrays instead of pointers and save the data there

#define IMAGE_SIZE 1256 //Put here the actual size of your image data
struct cap_data {
	guint8[IMAGE_SIZE] cap_data_a;
	guint8[IMAGE_SIZE] cap_data_a;
	// ...
};

However, as said, it's not recommended to send plain structs. However, if you are tied to C++, eCAL does not support any serialization formats on top of this. There is no chance for you to use C++ on the sender side?

**Comment by piyushmz** [comment]
Hello @KerstinKeller  I've a jetson orin Ubuntu 20.04 & 6 GMSL cameras  are connected to it. What I need is I want frame from each camera and concatenate them together in a buffer & this buffer is to be (sent)/ included in my perception module. Can the ecal_cam_samples do it or what are other ways I can do. If any other ideas you have please do suggest. I'm trying [jetson-utils](https://github.com/dusty-nv/jetson-utils/tree/master/camera) which is giving out a single camera o/p. I've a gstreamer camera driver in C which is already discussed.

**Comment by FlorianReimold** [comment]
Hi @piyushmz,

tl;dr: No, this sample cannot concatenate frames

---

The eCAL Camera Sample uses the Qt Library to access a video device. On Linux, those are usually mounted to `/dev/video0` , `/dev/video1` etc. On Windows the mount points are more difficult to determine, but I see you are using Linux anyways.
The sample then grabs an image in `jpg` compression from the camera and puts the frame into a protobuf message. you can find it here: https://github.com/eclipse-ecal/ecal-camera-samples/blob/main/proto_messages/compressed_image.proto. That frame is then sent out to a certain eCAL Topic and can be viewed with the eCAL Monitor Plugin.

This repository is meant as a sample application that can be used to demonstrate the functionality. You can also use it as a starting  point to develop your own specific application. However, as you want to use Raw C and this sample is written in C++, it will likely not help you.

You will have to write your own C application, that:
- Grabs all of your cameras
- Puts the data in contiguous memory
- sends out the contiguous memory as eCAL message, using traditional `(void* data, int length)`.

--------

Addition: I don't suggest you should concatenate camera frames to one giant binary blob. On the contrary, eCAL will work better with individual frames either published to the same topic or to 6 different topics. So if concatenated frames isn't a hard requirement, I strongly advise against that.

**Comment by piyushmz** [comment]
Thanks for your help.
